packages/context/src/handler-chain.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {BindingFilter} from './binding-filter';
+import {BindingAddress} from './binding-key';
+import {BindingComparator} from './binding-sorter';
+import {Context} from './context';
+import {transformValueOrPromise, ValueOrPromise} from './value-promise';
+
+const debug = debugFactory('loopback:context:handler-chain');
+
+/**
+ * Return value for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationResult = any;
+
+/**
+ * Array of arguments for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationArgs = any[];
+
+/**
+ * `next` function
+ */
+export type Next = () => ValueOrPromise<InvocationResult>;
+
+/**
+ * A handler function to be invoked in a chain for the given context
+ */
+export interface Handler<C extends Context = Context> {
+  /**
+   * @param context - Context object
+   * @param next - A function to proceed with downstream handlers or the target
+   *
+   * @returns The invocation result as a value or promise
+   */
+  (context: C, next: Next): ValueOrPromise<InvocationResult>;
+}
+
+/**
+ * Handler function or a binding key that resolves a handler function
+ */
+export type HandlerOrKey<C extends Context = Context> =
+  | BindingAddress<Handler<C>>
+  | Handler<C>;
+
+/**
+ * Invocation state of a handler chain
+ */
+class HandlerChainState<C extends Context = Context> {
+  private _index: number = 0;
+  /**
+   * Create a state for the handler chain
+   * @param handlers - Handler functions or binding keys
+   */
+  constructor(private handlers: HandlerOrKey<C>[]) {}
+
+  get index() {
+    return this._index;
+  }
+
+  /**
+   * Check if the chain is done - all handlers are invoked
+   */
+  done() {
+    return this._index === this.handlers.length;
+  }
+
+  /**
+   * Get the next handler to be invoked
+   */
+  next() {
+    if (this.done()) {
+      throw new Error('No more handler is in the chain');
+    }
+    return this.handlers[this._index++];
+  }
+}
+
+/**
+ * A chain of handlers to be invoked for the given context
+ */
+export class InvocationHandlerChain<C extends Context = Context> {
```

Comment:
If we use `Middleware` instead of `Handler,` then I think this can be renamed to `MiddlewareChain`? That reads much better to me than `InvocationHandlerChain`.

Reply:
If we end up with `Handler` as the type name, would it make sense to shorten this class name to `HandlerChain`? I don't understand the `Invocation` part.